File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        res = []

        def dfs(curr):
            if not curr:
                res.append("N")
                return
            res.append(str(curr.val))
            dfs(curr.left)
            dfs(curr.right)
        dfs(root)
        return ",".join(res)


        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        if data == "":
            return None 
        data = data.split(",")
        i = 0
        def dfs():
            nonlocal i
            if i >= len(data):
                return None
            if data[i] == "N":
                i += 1
                return None
            
            node = TreeNode(int(data[i]))
            i += 1
            node.left = dfs()
            node.right = dfs()
            return node
        
        return dfs()

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    boolean found = false; 

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        List<TreeNode> path1 = new ArrayList<>();
        List<TreeNode> path2 = new ArrayList<>();
        dfs(root, path1, p);
        found = false;
        dfs(root, path2, q);

        for (int i = Math.min(path1.size(), path2.size()) - 1; i >= 0; i --){
            if (path1.get(i) == path2.get(i)){
                return path1.get(i);
            }
        }

        return null;
    }

    private void dfs(TreeNode node, List<TreeNode> path, TreeNode target) {
        if (found) {
            return;
        }

        if (node == null) {
            return;
        }

        path.add(node);
        if (node == target) {
            found = true;
            return;
        }

        dfs(node.left, path, target);
        dfs(node.right, path, target);
        if (found) {
            return;
        }
        path.remove(path.size() - 1);
    }
}
